### PR TITLE
SubLevel 8 | Assignment 16 |  IDisposable and Dispose Pattern

### PR DESCRIPTION
**Pull request by shafiq005-solin**
# Description
This assignment is for the understanding of the managed and unmanaged resources and how the I Disposable were helping in the unmanaged resources. Why dispose method is used compare to the finalizer.

# How Has This Been Tested?

- [x] Managed resources
- [x] Unmanaged resources

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿// <copyright file="ManagedResource.cs" company="PlaceholderCompany">
+// Copyright (c) PlaceholderCompany. All rights reserved.
+// </copyright>
+
+namespace IDisposableAndDisposePattern
+{
+    /// <summary>
+    /// Manages the Resources with primary methods
```
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs:8]: Remove unwanted capitalization in all occurences.

File: src/Assignment-16/IDisposableAndDisposePattern/UnManagedResource.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            Console.WriteLine("UnManagedResource finalizer were called");
+        }
+
+        /// <summary>
+        /// Gets or Sets the Writer for the resource
+        /// </summary>
+        /// <value> StreamWriter </value>
+        public StreamWriter Writer { get; set; }
+
+        private static bool IsDisposed { get; set; } = false;
```
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/UnManagedResource.cs:41]: private variable should be at the top of the class  and should start with underscore.

Comment (MithunSubramani) [reply src/Assignment-16/IDisposableAndDisposePattern/UnManagedResource.cs:41]: false should be updated in constructor, right?

File: src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            this.Data.Add(item);
+        }
+
+        /// <summary>
+        /// Clear the Data
+        /// </summary>
+        public void ClearData()
+        {
+            this.Data.Clear();
+            IsDisposed = true;
```
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs:57]: Pass it as a parameter to dispose method

File: src/Assignment-16/IDisposableAndDisposePattern/Program.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    /// </summary>
+    internal class Program
+    {
+        /// <summary>
+        /// Main method for the <see cref="IDisposableAndDisposePattern"/> namespace
+        /// </summary>
+        /// <param name="args"> Console arguments </param>
+        public static void Main(string[] args)
+        {
+            Fn();
```
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/Program.cs:18]: Rename it to a meaningful name.

File: src/Assignment-16/IDisposableAndDisposePattern/Program.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            // unManaged.CloseFile();
+            // unManaged.WriteToFile("This will throw an exception!"); Throws error.
+            ManagedResource managed1 = new ManagedResource();
+            Console.WriteLine("\n\nTom was added to the ManagedResource.");
+            managed1.AddData("Tom");
+            Console.WriteLine("ManagedResouce contains : " + managed1.Data[0]);
+
+            UnManagedResource unManaged1 = new UnManagedResource(Path.Combine(directoryInfo.Parent?.Parent?.Parent?.FullName!, "text1.txt"));
+            Console.WriteLine("\n\nAnny was added to the UnManagedResource.");
+            unManaged1.WriteToFile("Anny");
```
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/Program.cs:49]: Clean up the code

File: src/Assignment-16/IDisposableAndDisposePattern/readme.md
Comment (MithunSubramani) [inline src/Assignment-16/IDisposableAndDisposePattern/readme.md:1]: Rephrase the document removing the pronouns

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        [Fact]
+        public void Input_AddDate_InputAddedInList()
+        {
+            ManagedResource managedResource = new ManagedResource();
+            List<string> list = new List<string>();
+            list.Add("one");
+            
+            managedResource.AddData("one");
+
+            Assert.Equal(list, managedResource.Data);
```
Comment (MithunSubramani) [inline tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:16]:
```suggestion
            List<string> list = new List<string>();
            list.Add("one");
            
            ManagedResource managedResource = new ManagedResource();
            managedResource.AddData("one");

            Assert.Equal(list, managedResource.Data);
```

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        [Fact]
+        public void Input_WriteToFile_InputWasWrittenInFile()
+        {
+            UnManagedResource unManagedResource = new UnManagedResource($"{Directory.GetCurrentDirectory}+./../../../Test.txt");
+
+            unManagedResource.WriteToFile("one");
+            unManagedResource.CloseFile();
+            StreamReader reader = new StreamReader($"{Directory.GetCurrentDirectory}+./../../../Test.txt");
+            string actual = reader.ReadToEnd();
```
Comment (MithunSubramani) [inline tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:27]: `actual??` Please update the variable name

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
+using IDisposableAndDisposePattern;
+
+namespace IDisposableAndDisposePatternTests
+{
+    public class UnitTest1
+    {
+        [Fact]
+        public void Input_AddDate_InputAddedInList()
```
Comment (MithunSubramani) [inline tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:8]:
```suggestion
        public void Input_AddData_InputAddedInList()
```

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs
Comment (MithunSubramani) [inline tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:1]: Add failure test cases too.

Comment (MithunSubramani) [inline tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:1]: Rename the test file name.

**Review by MithunSubramani — CHANGES_REQUESTED** [review]

File: src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs (hunk 1 above)
Comment (shafiq005-solin) [reply src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs:8]: All the unwanted capitalization were removed.

File: src/Assignment-16/IDisposableAndDisposePattern/UnManagedResource.cs (hunk 2 above)
Comment (shafiq005-solin) [reply src/Assignment-16/IDisposableAndDisposePattern/UnManagedResource.cs:41]: The template shows error when we use an underscore in private property. It is a static variable so, there is no need to initialize this variable in constructor.

File: src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs (hunk 3 above)
Comment (shafiq005-solin) [reply src/Assignment-16/IDisposableAndDisposePattern/ManagedResource.cs:57]: Parameter was passed to the method.

File: src/Assignment-16/IDisposableAndDisposePattern/Program.cs (hunk 5 above)
Comment (shafiq005-solin) [reply src/Assignment-16/IDisposableAndDisposePattern/Program.cs:49]: This set of code is directly copied from the question, we have to ensure that this set of code should work as excepted. And I add some extra comments in the code to understand easier.

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs (hunk 6 above)
Comment (shafiq005-solin) [reply tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:16]: format updated

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs (hunk 7 above)
Comment (shafiq005-solin) [reply tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:27]: Variable name were changed

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs (hunk 8 above)
Comment (shafiq005-solin) [reply tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:8]: Method name were updated.

File: tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs
Comment (shafiq005-solin) [reply tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:1]: Failure test cases also added.

Comment (shafiq005-solin) [reply tests/unit/Assignment-16/IDisposableAndDisposePatternTests/UnitTest1.cs:1]: Test file name was renamed.

File: src/Assignment-16/IDisposableAndDisposePattern/readme.md
Comment (shafiq005-solin) [reply src/Assignment-16/IDisposableAndDisposePattern/readme.md:1]: The document was rephrased.

**Review by MithunSubramani — APPROVED** [review]

**Pull request merged**